internal/app/machined/pkg/controllers/runtime/pcirebind.go

```diff
@@ -0,0 +1,224 @@
+	vendorDeviceToBoundDriver        map[string]string
+}
+
+// Name implements controller.Controller interface.
+func (c *PCIRebindController) Name() string {
+	return "runtime.PCIRebindController"
+}
+
+// Inputs implements controller.Controller interface.
+func (c *PCIRebindController) Inputs() []controller.Input {
+	return nil
+}
+
+// Outputs implements controller.Controller interface.
+func (c *PCIRebindController) Outputs() []controller.Output {
+	return []controller.Output{
+		{
+			Type: runtime.PCIRebindStatusType,
+			Kind: controller.OutputExclusive,
+		},
+	}
+}
+
+// Run implements controller.Controller interface.
+//
+//nolint:gocyclo,cyclop
+func (c *PCIRebindController) Run(ctx context.Context, r controller.Runtime, logger *zap.Logger) (err error) {
+	if c.initialVendorDeviceToBoundDriver == nil {
+		c.initialVendorDeviceToBoundDriver = make(map[string]string)
+	}
+
+	if c.vendorDeviceToBoundDriver == nil {
+		c.vendorDeviceToBoundDriver = make(map[string]string)
+	}
+
+	// wait for udevd to be healthy,
+	if err := WaitForDevicesReady(ctx, r,
+		[]controller.Input{
+			{
+				Namespace: runtime.NamespaceName,
+				Type:      runtime.PCIRebindConfigType,
+				Kind:      controller.InputWeak,
+			},
+		}); err != nil {
+		return fmt.Errorf("error waiting for devices to be ready: %w", err)
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-r.EventCh():
+		}
+
+		// Skip PCI rebind handling if running in a container or agent mode.
+		if c.V1Alpha1Mode.InContainer() || c.V1Alpha1Mode.IsAgent() {
+			return nil
+		}
+
+		pciRebindConfigs, err := safe.ReaderListAll[*runtime.PCIRebindConfig](ctx, r)
+		if err != nil {
+			return fmt.Errorf("error listing all PCI rebind configs: %w", err)
+		}
+
+		r.StartTrackingOutputs()
+
+		for cfg := range pciRebindConfigs.All() {
+			if _, ok := c.initialVendorDeviceToBoundDriver[cfg.TypedSpec().VendorDeviceID]; !ok {
+				boundDriver, err := checkDeviceBoundDriver(cfg.TypedSpec().VendorDeviceID)
+				if err != nil {
+					return fmt.Errorf("error checking already bound driver for device with id: %s, %w", cfg.TypedSpec().VendorDeviceID, err)
+				}
+
+				c.initialVendorDeviceToBoundDriver[cfg.TypedSpec().VendorDeviceID] = boundDriver
+			}
+		}
+
+		touchedIDs := map[string]struct{}{}
+
+		for cfg := range pciRebindConfigs.All() {
+			if targetDriver, ok := c.vendorDeviceToBoundDriver[cfg.TypedSpec().VendorDeviceID]; !ok || targetDriver != cfg.TypedSpec().TargetDriver {
+				if err := c.handlePCIBindToTarget(cfg.TypedSpec().VendorDeviceID, cfg.TypedSpec().TargetDriver); err != nil {
+					return err
+				}
+
+				boundDriver, err := checkDeviceBoundDriver(cfg.TypedSpec().VendorDeviceID)
+				if err != nil {
+					return fmt.Errorf("error checking bound driver for device with id: %s, %w", cfg.TypedSpec().VendorDeviceID, err)
+				}
+
+				if boundDriver != cfg.TypedSpec().TargetDriver { // TODO: this will fail if the module is not loaded
```

Comment:
should we add a module status and see if it exists?

Reply:
Maybe the pcirebind controler could create kernel module configs (automatically request the module to be loaded when user wants to rebind to it), and consume the status from module controller?